Jenkinsfile

```diff
@@ -0,0 +1,147 @@
+        apiToken = env.GITHUB_API_PASSWORD
+      }
+      gitHub = new GitHub(this, "${organization}/${repository}", apiToken)
+    }
+    stage('Build') {
+      gitHub.statusUpdate commitId, 'pending', 'build', 'Build is running'
+
+      def hash = OsTools.runSafe(this, "docker build --quiet --no-cache --tag ${imageName} .")
+      imageId = hash.split(':')[1]
+
+      if (currentBuild.result == 'FAILURE') {
+        gitHub.statusUpdate commitId, 'failure', 'build', 'Build failed'
+        return
+      } else {
+        gitHub.statusUpdate commitId, 'success', 'build', 'Build succeeded'
+      }
+    }
+    stage('Test') {
+      gitHub.statusUpdate commitId, 'pending', 'test', 'Tests are running'
+
+      def gemInstallDirectory = getGemInstallDirectory()
+      withEnv(["PATH+GEMS=${gemInstallDirectory}/bin"]) {
+        OsTools.runSafe(this, "gem install --user-install rspec")
+        OsTools.runSafe(this, "gem install --user-install serverspec")
+        OsTools.runSafe(this, "gem install --user-install docker-api")
+        OsTools.runSafe(this, "IMAGE_ID=${imageId} rspec --backtrace spec/Dockerfile_spec.rb")
+      }
+
+      if (currentBuild.result == 'FAILURE') {
+        gitHub.statusUpdate commitId, 'failure', 'test', 'Tests failed'
+        return
+      } else {
+        gitHub.statusUpdate commitId, 'success', 'test', 'Tests succeeded'
+      }
+    }
+    if (currentBuild.result == 'FAILURE') {
+      return
+    }
+    stage('Archive') {
+      dir('build/target') {
+        OsTools.runSafe(this, "docker save ${imageName} | gzip > ${archiveName}.tar.gz")
+        archiveArtifacts artifacts: "${archiveName}.tar.gz", onlyIfSuccessful: true
+      }
+    }
+    if (scm.branches[0].name != '*/master') {
+      return
+    }
+    input 'Push tags and image?'
+    stage('Push tags') {
+      withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: credentialsId,
+                        usernameVariable: 'GITHUB_API_USERNAME', passwordVariable: 'GITHUB_API_PASSWORD']]) {
+        OsTools.runSafe(this, "git tag ${version}")
+        OsTools.runSafe(this, """
+          git push \
+          https://${env.GITHUB_API_USERNAME}:${env.GITHUB_API_PASSWORD}@github.com/${organization}/${repository}.git \
+            ${version}
+        """)
+      }
+      OsTools.runSafe(this, "git tag -d ${version}")
+    }
+    stage('Push image') {
+      def dockerhubApiToken
+      withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: 'docker-hub-credentials',
+          usernameVariable: 'DOCKERHUB_API_USERNAME', passwordVariable: 'DOCKERHUB_API_PASSWORD']]) {
+        OsTools.runSafe(this, "docker tag ${imageId} ${organization}/${repository}:${version}")
+        OsTools.runSafe(this, "docker tag ${imageId} ${organization}/${repository}:latest")
+        OsTools.runSafe(this, """
+          echo ${env.DOCKERHUB_API_PASSWORD} | docker login --username ${env.DOCKERHUB_API_USERNAME} \
```

Comment:
IIUC the password is marked from the logs because this is within the `withCredentials` block. Is this true?

Reply:
Yes. The password and username are blocked out with *. Also, it appears that the version of docker running on the Jenkins slave does not support the --password-stdin option.

<img width="526" alt="screen shot 2017-10-18 at 2 52 13 pm" src="https://user-images.githubusercontent.com/865101/31744539-09592b48-b414-11e7-9c5f-527f15361dec.png">

Reply:
I will probably have to change this to use the --password option instead.

Reply:
The password will be blocked out with * as well.

Reply:
Here it is with the --password option:

<img width="411" alt="screen shot 2017-10-18 at 3 15 49 pm" src="https://user-images.githubusercontent.com/865101/31745554-1a07a812-b418-11e7-95c5-49502d1f37b1.png">